### DISCUSSION TITLE
@xstate/inspect : transition to child hist state causes error until page is refreshed

### DISCUSSION THREAD

**Discussion by JayNaire** [statelyai/xstate#1861] [Q&A] [answered]
EDIT: I suspect this is more serious than I thought as, in some circumstances, transitions to a child history state just do not happen at all.

Working gist is [here](https://xstate.js.org/viz/?gist=96ace7e4e00e2986fe734cffef7ff874)
Initial state is softOff.
Sending event "SwitchOn" in that gist correctly transitions to control.hist -> target -> control.total
However, if I parcel serve that machine so:
```...
   var serviceM = interpret(m, { devTools: true })
   serviceM.start()
...
```
sending the SwitchOn event (by inspect or programmatically) causes this error:
```
Uncaught TypeError: _utils.nestedPath(...)(...) is undefined
    resolveHistory StateNode.js:1298
    getRelativeStateNodes StateNode.js:1194
    allNextStateNodes StateNode.js:605
    next StateNode.js:604
    transitionLeafNode StateNode.js:435
    _transition StateNode.js:528
    transition StateNode.js:801
    nextState interpreter.js:695
    provide serviceScope.js:9
    nextState interpreter.js:694
    send interpreter.js:102
    process scheduler.js:64
    schedule scheduler.js:43
    send interpreter.js:98
    inspectSend browser.js:113
    actions inspectMachine.js:38
    exec interpreter.js:756
    execute interpreter.js:198
    update interpreter.js:224
    send interpreter.js:104
    process scheduler.js:64
    schedule scheduler.js:43
    send interpreter.js:98
    messageHandler browser.js:78
StateNode.js:1298:60 
```
**If I simply refresh the page** ( http://localhost:1234/ ) **everything works as expected.**

Why should this be?

**Comment by JayNaire** [comment, marked as answer]
SOLVED: Child states must have an explicit initial state specified.A history target is not sufficient.
```
...
      control: {
        initial: "total", <--- causes errors if omitted
        on: {
          // ,
          TICK:{actions:["TICK"]},
          Off: 'softOff',
          Service: "service",
          Total: "control.total",
          Ch: "control.ch",
          Hw: "control.hw",
          Frost: "control.frost",
          Flush: "control.flush",
          Bleed: "control.bleed",
          Balance: "control.balance",
          _brownout: "brownout",
          _fault: "fault"
        },
        states: {
          total: {},
          ch: {},
          hw: {},
          frost: {},
          flush: {},
          bleed: {},
          balance: {},
          hist: {
            type: "history",
            target: "total" <--- not sufficient as a default state in some transitions
          }
        }
      },
...
```

**Comment by Andarist** [comment]
I haven't fully dived into it - but looks that it might be a bug. Could you report your original issue with the history state as an issue? Preferably with a codesandbox repro case attached to it?

**Reply by davidkpiano** [reply]
Should we warn about compound states with missing initial targets?

**Reply by Andarist** [reply]
Definitely! Unless we want to pick some defaults (document order?) like in SCXML - we currently don't do that and this results in invalid configurations (talking about configuration in SCXML sense). A machine just has to resolve to leaf nodes, shouldn't be stuck at some compound level.

I would probably be against picking up a default though, it's a little bit magical and might be surprising to new learners. OTOH, not every compound needs the initial state to be defined but I guess this is rather rare - only compound states that are targets of transitions (directly or indirectly) need them to resolve to correct leaves. If one targets a lower-level state (through id or smth) the initial state might not be needed if the algorithm will never need to use it.

**Reply by davidkpiano** [reply]
Yeah, this is the dilemma: initial states aren't always necessary, but defaulting to the first one would be too magical.

~We can enforce that an `initial` state _must_ be specified only if the parent state is ever targeted directly~ Actually never mind; it would need `initial` if any ancestors are targeted directly either. Making `initial` required makes the most sense.

**Reply by Andarist** [reply]
Yep - the ancestor case is what ive meant by targeting a state indirectly/implicitly.

Making initial required at all times makes sense - if it gets proven that it's annoying for some cases we might always loosen the requirement a little bit but i doubt it will get to that
